File: src/main/scala/io/vertx/ext/asyncsql/impl/pool/PostgresqlAsyncConnectionPool.scala
package io.vertx.ext.asyncsql.impl.pool

import com.github.mauricio.async.db.Configuration
import com.github.mauricio.async.db.postgresql.PostgreSQLConnection
import io.netty.channel.EventLoop
import io.vertx.core.Vertx
import io.vertx.core.logging.Logger
import io.vertx.core.logging.impl.LoggerFactory

import scala.concurrent.ExecutionContext

class PostgresqlAsyncConnectionPool(val vertx: Vertx, config: Configuration, eventLoop: EventLoop, val maxPoolSize: Int) extends AsyncConnectionPool {

  private val logger: Logger = LoggerFactory.getLogger(classOf[PostgresqlAsyncConnectionPool])

  private implicit val executionContext: ExecutionContext = SimpleExecutionContext(logger)

  override def create() = new PostgreSQLConnection(
    configuration = config,
    group = eventLoop,
    executionContext = executionContext
  ).connect

}

object PostgresqlAsyncConnectionPool {
  def apply(vertx: Vertx, config: Configuration, eventLoop: EventLoop, maxPoolSize: Int): PostgresqlAsyncConnectionPool =
    new PostgresqlAsyncConnectionPool(vertx, config, eventLoop, maxPoolSize)
}

File: src/main/java/io/vertx/ext/asyncsql/impl/pool/PostgresqlAsyncConnectionPool.java
/*
 *  Copyright 2015 Red Hat, Inc.
 *
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the Eclipse Public License v1.0
 *  and Apache License v2.0 which accompanies this distribution.
 *
 *  The Eclipse Public License is available at
 *  http://www.eclipse.org/legal/epl-v10.html
 *
 *  The Apache License v2.0 is available at
 *  http://www.opensource.org/licenses/apache2.0.php
 *
 *  You may elect to redistribute this code under either of these licenses.
 */

package io.vertx.ext.asyncsql.impl.pool;

import com.github.mauricio.async.db.Configuration;
import com.github.mauricio.async.db.Connection;
import com.github.mauricio.async.db.postgresql.PostgreSQLConnection;
import com.github.mauricio.async.db.postgresql.column.PostgreSQLColumnDecoderRegistry;
import com.github.mauricio.async.db.postgresql.column.PostgreSQLColumnEncoderRegistry;
import io.vertx.core.Context;
import io.vertx.core.Vertx;

/**
 * Implementation of the {@link AsyncConnectionPool} for PostGresSQL.
 *
 * @author <a href="http://escoffier.me">Clement Escoffier</a>
 */
public class PostgresqlAsyncConnectionPool extends AsyncConnectionPool {

  public PostgresqlAsyncConnectionPool(Vertx vertx, int maxPoolSize, Configuration configuration) {
    super(vertx, maxPoolSize, configuration);
  }

  @Override
  protected Connection create() {
    return new PostgreSQLConnection(
        configuration,
        PostgreSQLColumnEncoderRegistry.Instance(),
        PostgreSQLColumnDecoderRegistry.Instance(),
        vertx.nettyEventLoopGroup().next(),
        executionContext);
  }

}
